File: frontend/src/services/api/plugin-dev.ts
import axios from './axios'
import { createEventStream } from './utils/stream'

const encodePathParam = (filePath: string): string => filePath.split('/').map(encodeURIComponent).join('/')

export type PluginDevTaskStatus =
  | 'pending'
  | 'running_cc'
  | 'creating_proposal'
  | 'waiting_apply'
  | 'applied'
  | 'failed'
  | 'cancelled'

export interface PluginDevVersionInfo {
  schema_version: number
  nekro_agent_channel: 'stable' | 'preview'
  nekro_agent_release: string
  nekro_agent_git_commit: string
  source_origin: 'runtime_snapshot' | 'cached_runtime' | 'remote_git' | 'cached_remote' | 'disabled' | 'unavailable'
  source_repo_url: string
  source_ref: string
  source_resolved_commit: string
  source_path: string
  source_dirty: boolean
  source_locked_at: string
  plugin_api_version: string
  stable_plugin_api_version: string
  template_version: string
  updated_at: string
  notes: string
}

export interface PluginDevStatusResponse {
  enabled: boolean
  sandbox_status: 'not_implemented' | 'stopped' | 'running' | 'failed'
  active_task_id: string | null
  queue_length: number
  cc_model_preset_id: number | null
  cc_model_preset_name: string | null
  version: PluginDevVersionInfo
}

export interface PluginDevGenerateRequest {
  file_path: string
  prompt: string
  current_code: string
  base_code: string
  dirty: boolean
  mode?: 'proposal'
}

export interface PluginDevGenerateResponse {
  task_id: string
  status: PluginDevTaskStatus
  proposal_id: string | null
}

export interface PluginDevTaskResponse {
  task_id: string
  file_path: string
  status: PluginDevTaskStatus
  summary: string
  logs: string[]
  proposal_id: string | null
  diff: string
  result_code: string
  error: string
  version: PluginDevVersionInfo
}

export interface PluginDevProposalResponse {
  proposal_id: string
  task_id: string
  file_path: string
  status: 'pending' | 'applied' | 'discarded'
  diff: string
  result_code: string
  summary: string
  created_at: string
}

export interface PluginDevApplyResponse {
  ok: boolean
  version_id: string
}

export interface PluginDevHistoryItem {
  version_id: string
  task_id: string
  action: string
  before_sha256: string
  after_sha256: string
  plugin_api_version: string
  nekro_agent_git_commit: string
  created_at: string
  summary: string
}

export interface PluginDevHistoryResponse {
  file_path: string
  current_version_id: string | null
  versions: PluginDevHistoryItem[]
}

export type PluginDevTaskStreamEvent =
  | { type: 'task'; task: PluginDevTaskResponse }
  | { type: 'done'; status: PluginDevTaskStatus }

export const pluginDevApi = {
  getStatus: async (): Promise<PluginDevStatusResponse> => {
    const response = await axios.get<PluginDevStatusResponse>('/plugin-dev/status')
    return response.data
  },

  start: async (): Promise<PluginDevStatusResponse> => {
    const response = await axios.post<PluginDevStatusResponse>('/plugin-dev/start')
    return response.data
  },

  stop: async (): Promise<PluginDevStatusResponse> => {
    const response = await axios.post<PluginDevStatusResponse>('/plugin-dev/stop')
    return response.data
  },

  getVersion: async (): Promise<PluginDevVersionInfo> => {
    const response = await axios.get<PluginDevVersionInfo>('/plugin-dev/version')
    return response.data
  },

  setCCModelPreset: async (presetId: number | null): Promise<PluginDevStatusResponse> => {
    const response = await axios.put<PluginDevStatusResponse>('/plugin-dev/cc-model-preset', {
      cc_model_preset_id: presetId,
    })
    return response.data
  },

  generate: async (body: PluginDevGenerateRequest): Promise<PluginDevGenerateResponse> => {
    const response = await axios.post<PluginDevGenerateResponse>('/plugin-dev/generate', body, {
      timeout: 120000,
    })
    return response.data
  },

  getTask: async (taskId: string): Promise<PluginDevTaskResponse> => {
    const response = await axios.get<PluginDevTaskResponse>(`/plugin-dev/tasks/${encodeURIComponent(taskId)}`)
    return response.data
  },

  cancelTask: async (taskId: string): Promise<PluginDevTaskResponse> => {
    const response = await axios.post<PluginDevTaskResponse>(`/plugin-dev/tasks/${encodeURIComponent(taskId)}/cancel`)
    return response.data
  },

  getProposal: async (proposalId: string): Promise<PluginDevProposalResponse> => {
    const response = await axios.get<PluginDevProposalResponse>(`/plugin-dev/proposals/${encodeURIComponent(proposalId)}`)
    return response.data
  },

  applyProposal: async (proposalId: string): Promise<PluginDevApplyResponse> => {
    const response = await axios.post<PluginDevApplyResponse>(`/plugin-dev/proposals/${encodeURIComponent(proposalId)}/apply`)
    return response.data
  },

  discardProposal: async (proposalId: string): Promise<void> => {
    await axios.delete(`/plugin-dev/proposals/${encodeURIComponent(proposalId)}`)
  },

  getHistory: async (filePath: string): Promise<PluginDevHistoryResponse> => {
    const response = await axios.get<PluginDevHistoryResponse>(`/plugin-dev/history/${encodePathParam(filePath)}`)
    return response.data
  },

  rollback: async (filePath: string, versionId: string, target: 'before' | 'after' = 'before'): Promise<PluginDevApplyResponse> => {
    const response = await axios.post<PluginDevApplyResponse>(`/plugin-dev/rollback/${encodePathParam(filePath)}`, {
      version_id: versionId,
      target,
    })
    return response.data
  },
}

export const streamPluginDevTask = (
  taskId: string,
  onEvent: (event: PluginDevTaskStreamEvent) => void,
  onError?: (error: Error) => void,
  signal?: AbortSignal
) => {
  return createEventStream({
    endpoint: `/plugin-dev/tasks/${encodeURIComponent(taskId)}/stream`,
    onMessage: data => {
      const trimmedData = data.trim()
      if (!trimmedData || !trimmedData.startsWith('{')) return
      try {
        const event = JSON.parse(trimmedData) as PluginDevTaskStreamEvent
        onEvent(event)
      } catch (error) {
        console.warn('Failed to parse plugin dev task stream event:', error, trimmedData)
      }
    },
    onError,
    signal,
  })
}

File: nekro_agent/schemas/plugin_dev.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

PluginDevTaskStatus = Literal[
    "pending",
    "running_cc",
    "creating_proposal",
    "waiting_apply",
    "applied",
    "failed",
    "cancelled",
]


class PluginDevVersionInfo(BaseModel):
    schema_version: int = 1
    nekro_agent_channel: Literal["stable", "preview"] = "preview"
    nekro_agent_release: str = ""
    nekro_agent_git_commit: str = ""
    source_origin: Literal[
        "runtime_snapshot",
        "cached_runtime",
        "remote_git",
        "cached_remote",
        "disabled",
        "unavailable",
    ] = "runtime_snapshot"
    source_repo_url: str = ""
    source_ref: str = ""
    source_resolved_commit: str = ""
    source_path: str = ""
    source_dirty: bool = False
    source_locked_at: str = ""
    plugin_api_version: str = "preview"
    stable_plugin_api_version: str = "stable"
    template_version: str = "1"
    updated_at: str
    notes: str = ""


class PluginDevStatusResponse(BaseModel):
    enabled: bool = True
    sandbox_status: Literal["stopped", "running", "failed"] = "stopped"
    active_task_id: str | None = None
    queue_length: int = 0
    cc_model_preset_id: int | None = None
    cc_model_preset_name: str | None = None
    version: PluginDevVersionInfo


class PluginDevCcModelPresetUpdate(BaseModel):
    cc_model_preset_id: int | None = None


class PluginDevVersionUpdate(BaseModel):
    nekro_agent_channel: Literal["stable", "preview"] | None = None
    nekro_agent_release: str | None = None
    nekro_agent_git_commit: str | None = None
    source_origin: Literal[
        "runtime_snapshot",
        "cached_runtime",
        "remote_git",
        "cached_remote",
        "disabled",
        "unavailable",
    ] | None = None
    source_repo_url: str | None = None
    source_ref: str | None = None
    source_resolved_commit: str | None = None
    source_path: str | None = None
    source_dirty: bool | None = None
    source_locked_at: str | None = None
    plugin_api_version: str | None = None
    stable_plugin_api_version: str | None = None
    template_version: str | None = None
    notes: str | None = None


class PluginDevGenerateRequest(BaseModel):
    file_path: str
    prompt: str = Field(..., min_length=1)
    current_code: str
    base_code: str = ""
    dirty: bool = False
    mode: Literal["proposal"] = "proposal"


class PluginDevInternalFileResponse(BaseModel):
    file_path: str
    content: str
    sha256: str


class PluginDevInternalFilePayload(BaseModel):
    file_path: str
    content: str


class PluginDevInternalProposalRequest(BaseModel):
    file_path: str
    content: str = Field(..., min_length=1)
    task_id: str = Field(default="plugin-dev-internal", min_length=1)
    summary: str = Field(default="由插件开发沙盒创建写入提案")
    files: list[PluginDevInternalFilePayload] = Field(default_factory=list)


class PluginDevInternalCheckRequest(BaseModel):
    file_path: str
    content: str = Field(..., min_length=1)
    task_id: str = Field(default="plugin-dev-internal", min_length=1)
    level: Literal["static", "load", "smoke", "strict"] = "static"
    files: list[PluginDevInternalFilePayload] = Field(default_factory=list)


class PluginDevGenerateResponse(BaseModel):
    task_id: str
    status: PluginDevTaskStatus
    proposal_id: str | None = None


class PluginDevTaskResponse(BaseModel):
    task_id: str
    file_path: str
    status: PluginDevTaskStatus
    summary: str = ""
    logs: list[str] = Field(default_factory=list)
    proposal_id: str | None = None
    diff: str = ""
    result_code: str = ""
    error: str = ""
    version: PluginDevVersionInfo


class PluginDevProposalFile(BaseModel):
    file_path: str
    content: str
    before_sha256: str = ""


class PluginDevProposalResponse(BaseModel):
    proposal_id: str
    task_id: str
    file_path: str
    status: Literal["pending", "applied", "discarded"]
    diff: str
    result_code: str
    summary: str
    created_at: str
    before_sha256: str = ""
    # 插件的完整文件集（含主文件）；为空表示旧版单文件提案，以 file_path/result_code 为准
    files: list[PluginDevProposalFile] = Field(default_factory=list)


class PluginDevApplyResponse(BaseModel):
    ok: bool = True
    version_id: str


class PluginDevHistoryItem(BaseModel):
    version_id: str
    task_id: str
    action: str
    before_sha256: str
    after_sha256: str
    plugin_api_version: str
    nekro_agent_git_commit: str
    created_at: str
    summary: str


class PluginDevHistoryResponse(BaseModel):
    file_path: str
    current_version_id: str | None = None
    versions: list[PluginDevHistoryItem] = Field(default_factory=list)


class PluginDevRollbackRequest(BaseModel):
    version_id: str
    target: Literal["before", "after"] = "before"


class PluginDevRollbackResponse(BaseModel):
    ok: bool = True
    version_id: str
